### PR TITLE
Do not use VS specific completion kinds in core LSP protocol layer

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes [AB#1777096](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1777096)

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 1]
```diff
@@ -79,7 +84,16 @@ internal static class ProtocolConversions
...
+        }.ToImmutableDictionary();
+
+        /// <summary>
+        /// Same as <see cref="RoslynTagToVSCompletionItemKind"/> except that the values are only from the safe set of
+        /// kinds defined directly by the core LSP protocol.
+        /// </summary>
+        public static readonly ImmutableDictionary<string, LSP.CompletionItemKind> RoslynTagToDefaultCompletionItemKind = RoslynTagToVSCompletionItemKind
+            .SetItem(WellKnownTags.Delegate, LSP.CompletionItemKind.Class)
+            .SetItem(WellKnownTags.ExtensionMethod, LSP.CompletionItemKind.Method)
+            .SetItem(WellKnownTags.Namespace, LSP.CompletionItemKind.Module);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:96]: picked what i thought were reasonable mappings.

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 2]
```diff
@@ -79,7 +84,16 @@ internal static class ProtocolConversions
...
             { WellKnownTags.NuGet, LSP.CompletionItemKind.Text }
-        };
+        }.ToImmutableDictionary();
+
+        /// <summary>
+        /// Same as <see cref="RoslynTagToVSCompletionItemKind"/> except that the values are only from the safe set of
+        /// kinds defined directly by the core LSP protocol.
+        /// </summary>
+        public static readonly ImmutableDictionary<string, LSP.CompletionItemKind> RoslynTagToDefaultCompletionItemKind = RoslynTagToVSCompletionItemKind
+            .SetItem(WellKnownTags.Delegate, LSP.CompletionItemKind.Class)
```
Comment (sharwell) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:94]: 💭 Maybe `Function` here?

**Review by sharwell — COMMENTED** [review]
This pull request should do one of the following:

1. Restrict to the known definitions in all cases
2. Restrict to the values provided by [CompletionClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionClientCapabilities)

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 3]
```diff
@@ -79,7 +84,16 @@ internal static class ProtocolConversions
...
             { WellKnownTags.AddReference, LSP.CompletionItemKind.Text },
             { WellKnownTags.NuGet, LSP.CompletionItemKind.Text }
-        };
+        }.ToImmutableDictionary();
+
+        /// <summary>
+        /// Same as <see cref="RoslynTagToVSCompletionItemKind"/> except that the values are only from the safe set of
+        /// kinds defined directly by the core LSP protocol.
+        /// </summary>
+        public static readonly ImmutableDictionary<string, LSP.CompletionItemKind> RoslynTagToDefaultCompletionItemKind = RoslynTagToVSCompletionItemKind
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:93]: I don't know what this means.  In VS we certainly would like our extension methods to appear as extension methods.

**Comment by CyrusNajmabadi** [comment]
> Restrict to the known definitions in all cases

I don't see any value in that.  This means we regress the experience in VS, despite it having extensions precisely so we will not.

> Restrict to the values provided by [CompletionClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionClientCapabilities)

That's what this PR attempts to implement.  If the client says it has VS extension capabilities, we use them.

**Comment by CyrusNajmabadi** [comment]
Ok.  We now use the kinds, if provided, to see what extra values we can return.

**Review by arkalyanms — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 4]
```diff
@@ -182,20 +185,35 @@ async Task<LSP.CompletionItem> CreateLSPCompletionItemAsync(CompletionItem item)
...
-                        return completionItemKind;
+                    if (ProtocolConversions.RoslynTagToCompletionItemKinds.TryGetValue(tag, out var completionItemKinds))
+                    {
+                        // Always at least pick the core kind provided.
+                        var kind = completionItemKinds[0];
+
+                        // If better kinds are preferred, return them if the client supports them.
+                        for (var i = 1; i < completionItemKinds.Length; i++)
+                        {
+                            var preferredKind = completionItemKinds[1];
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:210]:
```suggestion
                            var preferredKind = completionItemKinds[i];
```

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionTests.cs [hunk 5]
```diff
@@ -23,20 +23,24 @@ namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.Completion
 {
     public class CompletionTests : AbstractLanguageServerProtocolTests
     {
-        private static readonly LSP.VSInternalClientCapabilities s_vsCompletionCapabilities = new LSP.VSInternalClientCapabilities
-        {
-            SupportsVisualStudioExtensions = true,
-            TextDocument = new LSP.TextDocumentClientCapabilities
+        private static readonly LSP.VSInternalClientCapabilities s_vsCompletionCapabilities = CreateCoreCompletionCapabilities();
+
+        private static LSP.VSInternalClientCapabilities CreateCoreCompletionCapabilities ()
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionTests.cs:28]:
```suggestion
        private static LSP.VSInternalClientCapabilities CreateCoreCompletionCapabilities()
```

**Pull request merged**